### PR TITLE
[Improvement] Enhance progress bar

### LINKED ISSUE #58183: If the storage does not provide `total_rows_approx` for the progress bar, but it is a trivial query with LIMIT, provide `total_rows_approx` from the LIMIT.
**Use case**

This query should show progress bar:
```
INSERT INTO test SELECT * FROM generateRandom() LIMIT 100000000000;
```

### PR DESCRIPTION
**Pull request by lingtaolf**
Fixes: #58183

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Optimize the display of the progress bar in some cases when a `LIMIT` is provided.

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 50646539690b6ab54e28199ae4b471d3774b8a2e with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58935/50646539690b6ab54e28199ae4b471d3774b8a2e/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58935/50646539690b6ab54e28199ae4b471d3774b8a2e/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by rschu1ze** [comment]
@lingtaolf Which query did you use for testing? I tried

```sql
create table t (a String) engine=MergeTree order by tuple();
insert into t select * from generateRandom('a String') limit 100000000000;
```

but this PR did not make a difference for the progress bar.

**Comment by lingtaolf** [comment]
> ```sql
> ```sql
> insert into t select * from generateRandom('a String') limit 100000000000;
> ```
> 
@rschu1ze I used your sql , the progress bar can display on my termial. 

image

File: src/Parsers/ASTInsertQuery.h [hunk 1]
```diff
@@ -27,6 +27,7 @@ class ASTInsertQuery : public IAST
     ASTPtr watch;
     ASTPtr infile;
     ASTPtr compression;
+    UInt64 limit_value;
```
Comment (Algunenano) [inline src/Parsers/ASTInsertQuery.h:30]: Adding this here is not necessary. This logic does not belong to the parser.

File: src/Parsers/ParserInsertQuery.cpp [hunk 2]
```diff
@@ -168,6 +174,18 @@ bool ParserInsertQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
         pos = before_values;
         ParserSelectWithUnionQuery select_p;
         select_p.parse(pos, select, expected);
+        const ASTSelectWithUnionQuery ast_select = select->as<ASTSelectWithUnionQuery &>();
```
Comment (Algunenano) [inline src/Parsers/ParserInsertQuery.cpp:177]: This logic does not belong to the parser, instead it should go into the Interpreter.

**Review by Algunenano — CHANGES_REQUESTED** [review]

File: src/Interpreters/InterpreterInsertQuery.cpp [hunk 3]
```diff
@@ -391,9 +391,38 @@ Chain InterpreterInsertQuery::buildPreSinkChain(
...
 BlockIO InterpreterInsertQuery::execute()
 {
-    const Settings & settings = getContext()->getSettingsRef();
+    auto context_ptr = getContext();
+    const Settings & settings = context_ptr->getSettingsRef();
     auto & query = query_ptr->as<ASTInsertQuery &>();
 
+    if (query.select)
+    {
+        const ASTSelectWithUnionQuery * ast_select = typeid_cast<ASTSelectWithUnionQuery *>(query.select.get());
```
Comment (Algunenano) [inline src/Interpreters/InterpreterInsertQuery.cpp:400]: Why did you change it from using `->as<>` to cast?

It's what's used everywhere.

```c++
const auto & select_query = query.select->as<ASTSelectWithUnionQuery &>();
const ASTs & children = select_with_union->list_of_selects->children;
if (children.size() == 1)
...
```

Comment (lingtaolf) [reply src/Interpreters/InterpreterInsertQuery.cpp:400]: updated

Comment (Algunenano) [reply src/Interpreters/InterpreterInsertQuery.cpp:400]: It has crashed:

```
BaseDaemon: (version 24.1.1.1, build id: 0A7F232DDB35CBB660F88B6DCC0691852FB9112C, git hash: 2092bcbcc91012ad13b29d0759db504b37a806db) (from thread 153887) (query_id: 3c995a91-8ada-47db-b497-8720ffe84f29) (query: create table tab (x String) engine = MergeTree order by x as select 'Hello';) Received signal
```

File: src/IO/Progress.cpp [hunk 4]
```diff
@@ -5,6 +5,7 @@
 #include <IO/ReadHelpers.h>
 #include <IO/WriteHelpers.h>
 #include <Core/ProtocolDefines.h>
+#include <base/types.h>
```
Comment (Algunenano) [inline src/IO/Progress.cpp:8]: Unnecessary

**Review by Algunenano — COMMENTED** [review]
Please rebase with master and add a test.

File: tests/queries/0_stateless/02982_enhance_progress_bar.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
```
Comment (Algunenano) [inline tests/queries/0_stateless/02982_enhance_progress_bar.sh:1]: Test has many issues:
* no-parallel is not necessary
* Creating / deleting a database is not necessary (it's done automatically), just create the table
* `http://localhost:8123` is wrong, it should be `$CLICKHOUSE_URL`. There are many similar examples.

Comment (lingtaolf) [reply tests/queries/0_stateless/02982_enhance_progress_bar.sh:1]: Done

**Review by Algunenano — CHANGES_REQUESTED** [review]
Apart from the code being wrong, re-reading the issue and seeing the tests leads me into believing that this is a bad approach to solving the issue.

The current approach would only cover inserts and that's obviously wrong. We should not be trying to fix `INSERT`, we should be trying to fix the storage themselves so that it works both in any situation (select, insert, whatever)
